demos/source/aws_iot_demo_jobs.c

```diff
@@ -0,0 +1,533 @@
+
+/**
+ * @brief The MQTT Library timeout.
+ */
+#define TIMEOUT_MS ( 5000u )
+
+/**
+ * @brief The MQTT Library keepalive time.
+ */
+#define KEEP_ALIVE_SECONDS ( 60 )
+
+/**
+ * @brief The key of the Job ID.
+ */
+#define JOB_ID_KEY        "jobId"
+#define JOB_ID_KEY_LENGTH ( sizeof( JOB_ID_KEY ) - 1 )
+
+/**
+ * @brief Keys for the Job JSON response.
+ */
+#define JOB_DOC_KEY                  "jobDocument"
+#define JOB_DOC_KEY_LENGTH           ( sizeof( JOB_DOC_KEY ) - 1 )
+
+#define JOB_ACTION_KEY               "action"
+#define JOB_ACTION_KEY_LENGTH        ( sizeof( JOB_ACTION_KEY ) - 1 )
+
+#define JOB_MESSAGE_KEY              "message"
+#define JOB_MESSAGE_KEY_LENGTH       ( sizeof( JOB_MESSAGE_KEY ) - 1 )
+
+/**
+ * @brief Max Lengths for JSON values.
+ */
+#define JOBS_DEMO_MAX_ID_LENGTH      64
+#define JOBS_DEMO_MAX_JOB_DOC_LENGTH 64
+/**
+ * @brief The previous JobId and length
+ */
+static char _pJobId[ JOBS_DEMO_MAX_ID_LENGTH + 1 ] = {0};
+static size_t _JobIdLength = 0;
+
+/**
+ * @brief The Job Document which will be passed into the Notify Next callback.
+ */
+static char _jobDoc[ JOBS_DEMO_MAX_JOB_DOC_LENGTH + 1 ] = {0};
+static size_t _jobDocLength = 0;
+
+/**
+ * @brief Execute the Jobs Demo.
+ *
+ * This is the top level function called by the IOT Demo harness.
+ */
+int32_t RunJobsDemo( bool awsIotMqttMode,
+                     const char * pIdentifier,
+                     void * pNetworKServerInfo,
+                     void * pNetworkCredentialInfo,
+                     const IotNetworkInterface_t * pNetworkInterface );
+
+/**
+ * @brief Initialize The Jobs Demo.
+ *
+ * Initialize MQTT module then the Jobs module. If the Jobs module fails
+ * to initialize, MQTT cleanup is invoked.
+ *
+ * @return EXIT_SUCCESS if initialization succeeds. Otherwise EXIT_FAILURE.
+ */
+static int32_t _initDemo( void )
+{
+    int32_t status = EXIT_SUCCESS;
+    IotMqttError_t mqttInitStatus = IOT_MQTT_SUCCESS;
+    AwsIotJobsError_t jobsInitStatus = AWS_IOT_JOBS_SUCCESS;
+
+    mqttInitStatus = IotMqtt_Init();
+
+    if( mqttInitStatus != IOT_MQTT_SUCCESS )
+    {
+        status = EXIT_FAILURE;
+    }
+
+    if( status == EXIT_SUCCESS )
+    {
+        jobsInitStatus = AwsIotJobs_Init( AWS_IOT_DEMO_JOBS_MQTT_TIMEOUT );
+
+        if( jobsInitStatus != AWS_IOT_JOBS_SUCCESS )
+        {
+            IotMqtt_Cleanup();
+            status = EXIT_FAILURE;
+        }
+    }
+
+    return status;
+}
+
+/**
+ * @brief The callback when a new job is posted. It signals main via a semaphore.
+ */
+static void _jobsCallback( void * param1,
+                           AwsIotJobsCallbackParam_t * pCallbackInfo )
+{
+    bool success = false;
+    const char * pJsonValue = NULL;
+    size_t jsonValueLength = 0;
+    IotSemaphore_t * pWaitSem = ( IotSemaphore_t * ) param1;
+
+    IotLogInfo( "Job Doc Len : %d", pCallbackInfo->u.callback.documentLength );
+    IotLogInfo( "Job Document: %s", pCallbackInfo->u.callback.pDocument );
```

Comment:
Job document is not guaranteed to be NULL-terminated.  Must be printed with the length specifier.

Also, printing the length is probably not necessary.